trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/rest/RequestLoggerController.java

```diff
@@ -0,0 +1,211 @@
+    {
+        loggerProc = debugLogger;
+    }
+
+    public RequestLoggingSession newRequestSession(Request request, SigningServiceType serviceType)
+    {
+        return sessions.compute(request.requestId(), (requestId, current) -> {
+            checkState(current == null, "There is already a logging session for the request: " + requestId);
+            return internalNewRequestSession(request, serviceType);
+        });
+    }
+
+    public Optional<RequestLoggingSession> currentRequestSession(UUID requestId)
+    {
+        return Optional.ofNullable(sessions.get(requestId));
+    }
+
+    private RequestLoggingSession internalNewRequestSession(Request request, SigningServiceType serviceType)
+    {
+        if (!loggerProc.isEnabled()) {
+            return () -> {};
+        }
+
+        // TODO introduce telemetry/spans
+        UUID requestId = UUID.randomUUID();
+
+        Map<String, String> entries = new ConcurrentHashMap<>();
+
+        Map<String, String> properties = new ConcurrentHashMap<>();
+
+        Map<String, String> errors = new ConcurrentHashMap<>();
+
+        Map<String, Object> requestDetails = ImmutableMap.of(
+                "request.id", requestId,
+                "request.type", serviceType,
+                "request.uri", request.requestUri(),
+                "request.http.method", request.httpVerb(),
+                "request.http.entity", request.requestContent().contentType() != EMPTY);
+
+        addAll(entries, requestDetails);
+
+        logAndClear("RequestStart", entries);
+
+        return new RequestLoggingSession()
+        {
+            private final Stopwatch stopwatch = Stopwatch.createStarted();
+            private volatile boolean closed;
+
+            @Override
+            public void logProperty(String name, Object value)
+            {
+                properties.put(name, String.valueOf(value));
+            }
+
+            @Override
+            public void logError(String name, Object value)
+            {
+                errors.put(name, String.valueOf(value));
+            }
+
+            @SuppressWarnings({"ThrowableNotThrown", "SwitchStatementWithTooFewBranches"})
+            @Override
+            public void logException(Throwable e)
+            {
+                switch (Throwables.getRootCause(e)) {
+                    case WebApplicationException webApplicationException -> {
+                        errors.put("webException.status", Integer.toString(webApplicationException.getResponse().getStatus()));
+                        errors.put("webException.message", webApplicationException.getMessage());
+                    }
+
+                    default -> {
+                        errors.put("exception.type", e.getClass().getName());
+                        errors.put("exception.message", e.getMessage());
+                    }
+                }
+            }
+
+            @SuppressWarnings("resource")
+            @Override
+            public void close()
```

Comment:
I am a bit confused about what calls `close()` on a successful request. I saw that 2 error handler functions will call it.
There is also a [try with resource in the `StreamResponseHandler`](https://github.com/trinodb/aws-proxy/pull/82/files#diff-ced333cf1ae0df6f3f2c97d941bca937a5061d4e1709f472dd902aa7783d58ebR71), it that handling closing when no exceptions occur?

Reply:
Please see https://github.com/trinodb/aws-proxy/pull/83 which improves management of this.